tensorwaves/physics/helicityformalism/amplitude.py

```diff
@@ -1,246 +1,402 @@
+        for amp in self.seq_amp:
+            seq_amp = tf.multiply(seq_amp, amp(dataset))
         return seq_amp
 
 
-class _HelicityDecay(tf.keras.Model):
-    def __init__(self, recipe: dict):
-        super(_HelicityDecay, self).__init__(name="HelicityDecay")
+class _HelicityDecay:
+    def __init__(
+        self, recipe: dict, builder, kinematics, particle_list, dynamics
+    ):
         if isinstance(recipe, dict):
-            # call a register variable hook here that creates a new variable if it does not exist
-            # or if it exists simply returns that (thats how we can couple the parameters)
-            self.mass = tf.Variable(1.0, name="mass", dtype=tf.float64)
-            self.width = tf.Variable(0.1, name="width", dtype=tf.float64)
-            # also we should register kinematic variables here (with a string name)
-            # that will be handed down in the call method variable x.
-            # then we can do x[kin_var_name] to retrieve it
+            decaying_state = recipe["DecayParticle"]
+            decay_products = recipe["DecayProducts"]["Particle"]
+
+            def to_list(ids):
+                return ids if isinstance(ids, list) else [ids]
+
+            # define the subsystem
+            dec_prod_fs_ids = [x["FinalState"] for x in decay_products]
+            dec_prod_fs_ids = [to_list(x) for x in dec_prod_fs_ids]
+
+            recoil_fs_ids = []
+            parent_recoil_fs_ids = []
+            if "RecoilSystem" in recipe:
+                if "RecoilFinalState" in recipe["RecoilSystem"]:
+                    recoil_fs_ids = to_list(
+                        recipe["RecoilSystem"]["RecoilFinalState"]
+                    )
+                if "ParentRecoilFinalState" in recipe["RecoilSystem"]:
+                    parent_recoil_fs_ids = to_list(
+                        recipe["RecoilSystem"]["ParentRecoilFinalState"]
+                    )
+
+            (
+                self._inv_mass_name,
+                self._theta_name,
+                self._phi_name,
+            ) = kinematics.register_subsystem(
+                SubSystem(dec_prod_fs_ids, recoil_fs_ids, parent_recoil_fs_ids)
+            )
+
+            if decaying_state["Name"] not in particle_list:
+                raise LookupError(
+                    "Could not find particle with name "
+                    + decaying_state["Name"]
+                )
+            particle_infos = particle_list[decaying_state["Name"]]
+            decay_dynamics = [
+                x for x in dynamics if x["Name"] == decaying_state["Name"]
+            ]
+            if not decay_dynamics:
+                raise LookupError(
+                    "Could not find dynamics for particle with "
+                    "name " + decaying_state["Name"]
+                )
+            decay_dynamics = decay_dynamics[0]
+
+            non_dynamics_label = "nonDynamic"
+            known_dynamics_functions = {
```

Comment:
I wonder if we should move these kind of expandable dicts (implementations) to a move visible place, because I'm not sure how many steps need to be taken in order to implement a new dynamic function with this set-up.

Reply:
Yeah, I wanted to address this later for example in #44 .